src/main/kotlin/io/github/mojira/arisa/modules/privacy/PrivacyModule.kt

```diff
@@ -0,0 +1,242 @@
+import org.slf4j.LoggerFactory
+import java.nio.file.Files
+import java.time.Instant
+
+private fun Iterable<Regex>.anyMatches(string: String) = any { it.matches(string) }
+
+private val log = LoggerFactory.getLogger("PrivacyModule")
+
+class PrivacyModule(
+    private val message: String,
+    private val commentNote: String,
+    private val allowedEmailRegexes: List<Regex>,
+    private val sensitiveTextRegexes: List<Regex>,
+    private val attachmentRedactor: AttachmentRedactor,
+    private val sensitiveFileNameRegexes: List<Regex>
+) : Module {
+    // Matches an email address, which is not part of a user mention ([~name])
+    private val emailRegex = "(?<!\\[~)\\b[a-zA-Z0-9.\\-_]+@[a-zA-Z.\\-_]+\\.[a-zA-Z.\\-]{2,15}\\b".toRegex()
+
+    override fun invoke(issue: Issue, lastRun: Instant): Either<ModuleError, ModuleResponse> = with(issue) {
+        Either.fx {
+            assertNull(securityLevel).bind()
+
+            val (attachmentContainsSensitiveData, attachmentsToRedact) = checkAttachments(lastRun)
+            val issueContainsSensitiveData = attachmentContainsSensitiveData || containsIssueSensitiveData(lastRun)
+
+            val restrictCommentActions = getRestrictCommentActions(lastRun)
+
+            assertAny(
+                assertNotEmpty(attachmentsToRedact),
+                assertTrue(issueContainsSensitiveData),
+                assertNotEmpty(restrictCommentActions)
+            ).bind()
+
+            // Always try to redact attachments, even if issue would be made private anyways
+            // So in case issue was made private erroneously it can easily be made public
+            val redactedAll = redactAttachments(issue, attachmentsToRedact)
+
+            if (!redactedAll || issueContainsSensitiveData) {
+                setPrivate()
+                addComment(CommentOptions(message))
+            }
+
+            restrictCommentActions.forEach { it.invoke() }
+        }
+    }
+
+    private fun containsSensitiveData(string: String): TextRangeLocation? =
+        containsEmailMatch(string)
+            ?: sensitiveTextRegexes.asSequence()
+                .mapNotNull { it.find(string) }
+                .map { TextRangeLocation.fromMatchResult(string, it) }
+                .firstOrNull()
+
+    private fun containsEmailMatch(string: String): TextRangeLocation? {
+        return emailRegex
+            .findAll(string)
+            .filterNot { match -> allowedEmailRegexes.anyMatches(match.value) }
+            .map { TextRangeLocation.fromMatchResult(string, it) }
+            .firstOrNull()
+    }
+
+    private data class AttachmentsCheckResult(
+        /** Whether any of the attachments contains sensitive data which cannot be redacted */
+        val attachmentContainsSensitiveData: Boolean,
+        val attachmentsToRedact: List<RedactedAttachment>
+    )
+
+    private fun Issue.checkAttachments(lastRun: Instant): AttachmentsCheckResult {
+        var attachmentContainsSensitiveData: Boolean
```

Comment:
I feel like the name of this variable is not very good, I got quite confused reading this.

Reply:
Maybe something like `sensitiveDataFound`?

I guess the thing that confused me most is that this variable is not about a particular attachment, but rather about if _any_ attachment contains _any_ sensitive data.

Reply:
Changed it to `foundNonRedactableSensitiveData`; a bit verbose but hopefully describes its purpose well enough.